apis/run/v1alpha3/osimage_types.go

```diff
@@ -17,14 +19,38 @@ type OSInfo struct {
 }
 
 // MachineImageInfo describes the "Image" part of the OSImage, defined by the image type.
+// +kubebuilder:object:generate=false
 type MachineImageInfo struct {
 	// Type of the OSImage, roughly corresponding to the infrastructure provider (vSphere can serve both ova and vmop).
 	// Some of currently known types are: "ami", "azure", "docker", "ova", "vmop".
 	Type string `json:"type"`
 
 	// Ref is a key-value map identifying the image within the infrastructure provider. This is the data
 	// to be injected into the infra-Machine objects (like AWSMachine) on creation.
-	Ref map[string]string `json:"ref"`
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:validation:Type=object
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Ref map[string]interface{} `json:"ref"`
```

Comment:
Question: If the target is just an annotations on a runtime.Object, I am confused why this interface{}, is it because CRD machinery not able to marshal properly if it sees anything other string?

Reply:
> If the target is just an annotations on a runtime.Object, ...

Not sure I understand the above premise.

> why interface{}

Because we need to be able to preserve the data type, e.g.  `spec.image.marketplace.thirdPartyImage` field in AzureMachine is bool, not string.

If we keep the map value type as `string`, the infra-Machine webhook will need to be smart, which means less simple and more prone to bugs.

> is it because CRD machinery not able to marshal properly if it sees anything other string?

CRD machinery on the kube-apiserver side is perfectly capable of handling "free-form" `object` type. The added `+kubebuilder:...` annotations simply drive the CRD manifest generator to insert appropriate statements into the CRD YAML. 

Also, controller-gen cannot generate DeepCopy implementation `interface{}` on its own, so it had to be handled, too.

Reply:
non-blocking, approving for now